modality-probe-graph/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+//! `modality-probe-graph` is a library that builds different types of node
+//! and edge lists from a columnar, unordered collection of log
+//! reports like the one that modality-probe-udp-collector creates.
+use std::{collections::HashMap, hash::Hash};
+
+use err_derive::Error;
+
+use modality_probe::{EventId, LogicalClock, ProbeId};
+use modality_probe_collector_common::{EventLogEntry, Report, SequenceNumber};
+
+/// A trait for the inner graph type of `EventDiagraph`. This enables
+/// a custom inner graph that can be purpose built for a use-case, but
+/// allows said graph to still be built by `EventDigraph`.
+pub trait Graph {
+    /// Add a node / vertex to the inner graph.
+    fn add_node(&mut self, node: GraphEvent);
+    /// Add an edge to the inner graph.
+    fn add_edge(&mut self, source: GraphEvent, target: GraphEvent);
+}
+
+#[derive(Clone, Debug, Copy, PartialEq, Eq, Hash)]
+/// A node in the event digraph.
+pub struct GraphEvent {
+    pub id: EventId,
+    pub clock: LogicalClock,
+    pub payload: Option<u32>,
+    pub probe_id: ProbeId,
+    pub seq: SequenceNumber,
+    pub seq_idx: usize,
+}
+
+/// Errors returned by the `EventDigraph` methods.
+#[derive(Debug, Error)]
+pub enum Error {
+    /// Encountered an error when using the writer when exporting a
+    /// graph in a textual format.
+    #[error(display = "Formatting graph failed: {}", _0)]
+    Io(String),
+    /// The graph builders iterate over a `Result` to leave room for
+    /// deserialization; if a builder encounters a `Err`, this error
+    /// is returned with that error's `display` implementaion inside.
+    #[error(display = "An item in the log iterator was an error variant: {}", _0)]
+    ItemError(String),
+    /// Encountered an unexpected inconsistency in the graph data.
+    #[error(display = "Encountered an inconsistency in the graph data: {}", _0)]
+    InconsistentData(&'static str),
+}
+
+/// A type that eats Modality Probe reports and stuffs them into its
+/// inner graph type.
+#[derive(Debug, PartialEq, Eq)]
+pub struct EventDigraph<G: Graph> {
+    /// The EventDigraph retains ownership of the inner graph, but
+    /// must make it available so that its purpose-built functionality
+    /// is accessible.
+    pub graph: G,
+
+    // When a foreign trace clock is the last thing in a report chunk,
```

Comment:
I think using `///` is the right choice for fields, private or public. This allows switching field visibility without accidentally omitting the docstrings because the preceding lines didn't get an extra-slash added.